jerry-core/ecma/operations/ecma-arraybuffer-object.c

```diff
@@ -553,6 +555,156 @@ ecma_builtin_arraybuffer_slice (ecma_value_t this_arg, const ecma_value_t *argum
   return ret_value;
 } /* ecma_builtin_arraybuffer_slice */
 
+/**
+ * Apply atomics operation on numbers.
+ *
+ * Note: ECMA_ATOMICS_COMPAREEXCHANGE is not handled here.
+ *
+ * @return ecma value
+ */
+static ecma_value_t
+ecma_atomics_operation_on_number (ecma_value_t stored_value, ecma_value_t val, ecma_atomics_op_t op)
+{
+  ecma_number_t result = ecma_get_number_from_value (stored_value);
+
+  switch (op)
+  {
+    case ECMA_ATOMICS_ADD:
+    {
+      result += ecma_get_number_from_value (val);
+      break;
+    }
+    case ECMA_ATOMICS_AND:
+    {
+      ecma_value_t op_result = do_number_bitwise_logic (NUMBER_BITWISE_LOGIC_AND, stored_value, val);
+      result = ecma_get_number_from_value (op_result);
+      ecma_free_value (op_result);
+      break;
+    }
+    case ECMA_ATOMICS_EXCHANGE:
+    {
+      result = ecma_get_number_from_value (val);
+      break;
+    }
+    case ECMA_ATOMICS_OR:
+    {
+      ecma_value_t op_result = do_number_bitwise_logic (NUMBER_BITWISE_LOGIC_OR, stored_value, val);
+      result = ecma_get_number_from_value (op_result);
+      ecma_free_value (op_result);
+      break;
+    }
+    case ECMA_ATOMICS_SUBTRACT:
+    {
+      result -= ecma_get_number_from_value (val);
+      break;
+    }
+    default:
+    {
+      JERRY_ASSERT (op == ECMA_ATOMICS_XOR);
+      ecma_value_t op_result = do_number_bitwise_logic (NUMBER_BITWISE_LOGIC_XOR, stored_value, val);
+      result = ecma_get_number_from_value (op_result);
+      ecma_free_value (op_result);
+      break;
+    }
+  }
```

Comment:
```c
switch (op)
  {
    case ECMA_ATOMICS_ADD:
    {
      result += ecma_get_number_from_value (val);
      break;
    }
    case ECMA_ATOMICS_EXCHANGE:
    {
      result = ecma_get_number_from_value (val);
      break;
    }
    case ECMA_ATOMICS_SUBTRACT:
    {
      result -= ecma_get_number_from_value (val);
      break;
    }
    default:
    {
      ecma_value_t op_result = do_number_bitwise_logic ((number_bitwise_logic_op) (op - ECMA_ATOMICS_AND), stored_value, val);
      result = ecma_get_number_from_value (op_result);
      ecma_free_value (op_result);
      break;
    }
  }
```
Moreover, start the `ecma_atomics_op_t` enum with
```c
ECMA_ATOMICS_AND, /**< Atomics and operation */
  ECMA_ATOMICS_OR, /**< Atomics or operation */
  ECMA_ATOMICS_XOR, /**< Atomics xor operation */
```

and add static asserts to this file that `ECMA_ATOMICS_ADD == NUMBER_BITWISE_LOGIC_AND` and so on.